spec.bs

```diff
@@ -0,0 +1,1484 @@
+
+[=Ecosystems=] are helpful for use cases such as indexing the location of [=planted shape trees=].
+
+[=Shape tree=] client libraries should be initialized with a reference to the the target [=ecosystem=] it should operate within.
+
+For purposes of this specification, an [=ecosystem=] should be considered a software interface, expecting the below operations to be implemented.
+
+## Ecosystem Operations ## {#ecosystem-operations}
+
+### Get Existing Shape Tree ### {#ecosystem-getexistingshapetree}
+<code><b>getExistingShapeTreeFromContainer(context:ShapeTreeContext, parentContainer:URI, shapeTreesToPlant:ShapeTree[], resourceName:String) : ShapeTreeLocator</b></code>
+
+This operation allows the [=ecosystem=] to provide a previously [=planted shape tree=] location.
+
+### Before Plant Shape Tree ### {#ecosystem-beforeplantshapetree}
+<code><b>beforePlantShapeTree</b>(context:ShapeTreeContext, targetContainer:URI, bodyGraph:Graph, shapeTreesToPlant:ShapeTree[], linkHeaders:Map) : Graph</code>
+
+This operations allows the [=ecosystem=] to modify the the [=body graph=] of a [Plant Shape Tree](#plant-shapetree) operation request.
+
+### Index Shape Tree ### {#ecosystem-indexshapetree}
+<code><b>indexShapeTree</b>(context:ShapeTreeContext, parentContainer:URI, shapeTree:URI, targetContainer:URI) </code>
+
+This operation notifies the [=ecosystem=] of a new [=shape tree=] being planted.
+
+### Index Shape Tree Data Instance ### {#ecosystem-indexshapetreedatainstance}
+<code><b>indexShapeTreeDataInstance</b>(context:ShapeTreeContext, parentContainer:URI, shapeTree:URI, shapeTreeInstance:URI)</code>
+
+This operations notifies the [=ecosystem=] of a new [=shape tree data instance=] being created.
+
+### Unindex Shape Tree Data Instance ### {#ecosystem-unindexshapetreedatainstance}
+<code><b>unIndexShapeTreeDataInstance</b>(context:ShapeTreeContext, shapeTree:URI, shapeTreeInstance:URI)</code>
+
+This operation notifies the [=ecosystem=] of a [=shape tree data instance=] being deleted.
+
+### Unindex Shape Tree ### {#ecosystem-unindexshapetree}
+<code><b>unIndexShapeTree</b>(context:ShapeTreeContext, parentContainer:URI, shapeTree:URI, targetContainer:URI)</code>
+
+This operation notifies the [=ecosystem=] of a [=shape tree=]/[=managed container=] being deleted.
+
+Note: The <code><b>ShapeTreeContext</b></code> class is implementation specific and may be used to include important context such as authentication details, base URLs, etc.
+
+Shape Tree Operations {#operations}
+=====================
+
+Working with [=shape trees=] entails using several higher-level operations -- each of which may represent one or more HTTP requests and/or pieces of processing logic.
+
+The key operations used to manage [=shape trees=] are:
+* [Discover Shape Tree](#discover)
+* [Plant Shape Tree](#plant-shapetree)
```

Comment:
We should talk about this offline.  Specifically the pattern for if it would be updated in-place or at a different IRI.  If the definition is updated in place, it would be an "automatic" update -- good and bad.  I could see a use for an operation which would re-validate all existing resources to ensure conformance to some changed underlying definition.  But we'd have to determine the semantics of handling failing, existing resources.